src/Service/IdentityManagement/UserSession.php

```diff
@@ -443,4 +459,31 @@ private function outerPromise(PromiseInterface $inner): Promise
 
         return $outer;
     }
+
+    /**
+     * Sign in to MPX, with a lock to prevent sign-in stampedes.
+     *
+     * @param int $duration (optional) The number of seconds that the sign-in token should be valid for.
+     *
+     * @return Token
+     */
+    protected function signInWithLock(int $duration = null): Token
+    {
+        $factory = new Factory($this->store);
+        $factory->setLogger($this->logger);
+        $lock = $factory->createLock($this->user->getUsername(), 10);
```

Comment:
This creates a lock that lasts for 10 seconds. Note that any lock prefixing would be handled by decorating the store class that is passed in. The lock will release once `$lock` goes out of scope.